hash/hash_comparator.go

```diff
@@ -0,0 +1,107 @@
+package hash
+
+import (
+	"context"
+	"crypto/subtle"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/argon2"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/ory/kratos/driver/config"
+)
+
+var ErrUnknownHashAlgorithm = errors.New("unknown hash algorithm")
+
+func Compare(ctx context.Context, password []byte, hash []byte) error {
+	if isBcryptHash(hash) {
+		return CompareBcrypt(ctx, password, hash)
+	} else if isArgon2idHash(hash) {
+		return CompareArgon2id(ctx, password, hash)
+	} else {
+		return ErrUnknownHashAlgorithm
+	}
+}
+
+func CompareBcrypt(_ context.Context, password []byte, hash []byte) error {
+	if err := validateBcryptPasswordLength(password); err != nil {
+		return err
+	}
+
+	err := bcrypt.CompareHashAndPassword(hash, password)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CompareArgon2id(_ context.Context, password []byte, hash []byte) error {
+	// Extract the parameters, salt and derived key from the encoded password
+	// hash.
+	p, salt, hash, err := decodeArgon2idHash(string(hash))
+	if err != nil {
+		return err
+	}
+
+	// Derive the key from the other password using the same parameters.
+	otherHash := argon2.IDKey([]byte(password), salt, p.Iterations, p.Memory, p.Parallelism, p.KeyLength)
+
+	// Check that the contents of the hashed passwords are identical. Note
+	// that we are using the subtle.ConstantTimeCompare() function for this
+	// to help prevent timing attacks.
+	if subtle.ConstantTimeCompare(hash, otherHash) == 1 {
+		return nil
+	}
+	return ErrMismatchedHashAndPassword
+}
+
+func isBcryptHash(hash []byte) bool {
+	res, _ := regexp.Match("^\\$2[abzy]?\\$", hash)
+	return res
+}
+
+func isArgon2idHash(hash []byte) bool {
+	res, _ := regexp.Match("^\\$argon2id\\$", hash)
```

Comment:
Nitpicking: wouldn't `strings.HasPrefix()` be sufficient in this case?

Reply:
Agree, that could be a good simplification. Not sure if it is worth doing it now. I used this approach as it will allow adding more algorithms in the future and user regexes for identifying them.

Reply:
Nah. I think is not worth separate PR right now. When/if we tackle with #605 then that would be a good time to refactor this. 

Anyway @seremenko-wish thanks a lot for this PR, this will get us closer to get all critical features implemented. Is the #1175 closable now since hash comparison is fixed in this PR?

Reply:
closed #1175

this definitely opens gates for more algorithm integrations. Looks like PBKDF1/2 integration is missing, but that would be easy to add into existing hashing framework.

Reply:
Yup! Not sure if we want to support other algorithms as main hasher but definitely support them when importing users/hashes! This PR lays the foundation for doing that!